### PR TITLE
[GUI2] Selection and rendering of basic node widgets

### LINKED ISSUE #7732: Widget selection logic.

### PR DESCRIPTION
**Pull request by Frizi**
Fixes #7732

This PR adds a way to define widget components and automatically load them. So far only builtin ones are loaded, but that can be extended to a custom directory later, once the widget API is more fleshed out. The widgets now only support being created on specific AST elements. Changing that to allow widgets for "virtual elements", such as method argument placeholders, will be a separate task.

image

### CODE REVIEW

File: app/gui2/src/components/GraphEditor/GraphEdge.vue [hunk 1]
```diff
@@ -15,7 +16,8 @@ const props = defineProps<{
   edge: Edge
   nodeRects: Map<ExprId, Rect>
   exprRects: Map<ExprId, Rect>
-  exprNodes: Map<ExprId, ExprId>
+  lookupExprNode: (exprId: ExprId) => ExprId | undefined
+  lookupNodeColor: (nodeId: ExprId) => string | undefined
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/GraphEdge.vue]: This is a bit strange pattern: why cannot we just expose those methods from a common point? These functions are quite simple (and lookupNodeColor is actually an alias for graphStore.db method).

File: app/gui2/src/components/GraphEditor/GraphEdge.vue [hunk 2]
```diff
@@ -25,51 +27,64 @@ const emit = defineEmits<{
 
 const base = ref<SVGPathElement>()
 
-type PosMaybeSized = { pos: Vec2; size?: Vec2 }
+const sourceNode = computed(
+  () =>
+    props.edge.source ??
+    (selection?.hoveredNode != (props.edge.target && props.lookupExprNode(props.edge.target))
+      ? selection?.hoveredNode
+      : undefined),
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/GraphEdge.vue]: This is quite a long statement. Perhaps some named variables would make it clearer (I spent a while trying to understand why hovered node became the source node for anything).

File: app/gui2/src/components/GraphEditor/GraphEdge.vue [hunk 3]
```diff
@@ -25,51 +27,64 @@ const emit = defineEmits<{
...
-  sourceSize: Vec2 | undefined
+  sourceSize: Vec2
   /** The width and height of the port that the edge is attached to, if any. */
-  targetSize: Vec2 | undefined
+  targetSize: Vec2
   /** The coordinates of the node input port that is the edge's destination, relative to the source position.
    *  The edge enters the port from above. */
   targetOffset: Vec2
+  /** The distance between the target port top edge and the target node top edge. */
+  targetPortDistance: number | undefined
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/GraphEdge.vue]: Isn't this distance always the same? And basing on docs, this field should be named `targetPortTopMargin` or `targetPortTopDistanceInNode`.

Also, what if this is undefined? There is no distance?

File: app/gui2/src/components/GraphEditor/GraphNode.vue [hunk 4]
```diff
@@ -104,247 +106,41 @@ function getRelatedSpanOffset(domNode: globalThis.Node, domOffset: number): numb
...
-  } else if (document.caretRangeFromPoint) {
-    range = document.caretRangeFromPoint(event.clientX, event.clientY)
-    if (range == null) {
-      console.error('Could not find caret position when editing node.')
+function startEditingNode(position: Vec2 | undefined) {
+  let sourceOffset = 0
+  if (position != null) {
+    let domNode, domOffset
+    if ((document as any).caretPositionFromPoint) {
+      const range = document.caretPositionFromPoint(position.x, position.y)
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/GraphNode.vue]: Why is the variable called `range`?

File: app/gui2/src/components/GraphEditor/NodeWidget.vue [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+const filteredRootNode = computed(() => {
+  const element = rootNode.value
+  if (element instanceof HTMLElement) return element
+  else if (element?.$el instanceof HTMLElement) return element.$el
+  else return null
+})
+const exprRect = shallowRef<Rect>()
+const nodeSize = useResizeObserver(filteredRootNode, false)
+function updateRect() {
+  if (!isHoverable()) return
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/NodeWidget.vue]: So the exprRect will be not updated if widget is not hoverable? Perhaps we should warn about it next to `exprRect` definition?

File: app/gui2/src/components/GraphEditor/widgets/index.ts
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/widgets/index.ts:1]: Is this file needed?

File: app/gui2/src/components/GraphEditor/widgets/WidgetPort.vue [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  beforeOverride: true,
+  priority: 1,
+  match: (info) => {
+    const isConnected = info.db.connections.reverseLookup(info.ast.astId).size > 0
+    if (isConnected) {
+      return Score.Perfect
+    }
+    return Score.Mismatch
+  },
+})
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/widgets/WidgetPort.vue]: Do I understand correctly, that widget port is only a _connected_ port? It's intuitive. Perhaps rename it to "connected port"?

File: app/gui2/src/components/widgets/SliderWidget.vue [hunk 7]
```diff
@@ -27,24 +32,76 @@ const inputValue = computed({
     return props.modelValue
   },
   set(value) {
-    emit('update:modelValue', value)
+    if (typeof value === 'string') {
+      console.log(value, '->', toNumericOnly(value))
```
Comment (farmaazon) [inline app/gui2/src/components/widgets/SliderWidget.vue]: Stray log.

File: app/gui2/src/providers/widgetTree.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+export { injectFn as injectWidgetTree, provideFn as provideWidgetTree }
+const { provideFn, injectFn } = createContextStore(
+  'Widget tree',
+  (nodeSpanStart: Ref<number>) => new WidgetTree(nodeSpanStart),
+)
+
+export class WidgetTree {
+  constructor(public nodeSpanStart: Ref<number>) {}
+}
```
Comment (farmaazon) [inline app/gui2/src/providers/widgetTree.ts]: Do we expect to have more logic here? Can we write it in TODO?

File: app/gui2/src/providers/widgetRegistry.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+      if (alreadyUsed && alreadyUsed.has(module.default)) continue
+      const score = module.widgetConfig.match(info)
+      if (score === Score.Perfect) return module.default
+      if (score > bestScore) {
+        bestScore = score
+        best = module.default
+      }
+    }
+    return best
+  }
```
Comment (farmaazon) [inline app/gui2/src/providers/widgetRegistry.ts]: It would be nice to have tests for this logic, especially because it's about to get complex.

**Review by farmaazon — COMMENTED** [review]
First set of comments

File: app/gui2/src/providers/widgetUsageInfo.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
+import type { AstExtended } from '@/util/ast'
+import { computed, proxyRefs, type Ref } from 'vue'
+import { createContextStore } from '.'
+import type { WidgetComponent } from './widgetRegistry'
+
+export { injectFn as injectWidgetUsageInfo, provideFn as provideWidgetUsageInfo }
+const { provideFn, injectFn } = createContextStore(
+  'Widget usage info',
```
Comment (farmaazon) [inline app/gui2/src/providers/widgetUsageInfo.ts]: Please add docs. I don't quite get what is what in this. Is this "usage places of widgets" or "widgets used for given ast"?

File: app/gui2/src/stores/suggestionDatabase/index.ts [hunk 11]
```diff
@@ -85,7 +86,18 @@ class Synchronizer {
   private setupUpdateHandler(lsRpc: LanguageServer) {
     lsRpc.on('search/suggestionsDatabaseUpdates', (param) => {
       this.queue.pushTask(async ({ currentVersion }) => {
-        if (param.currentVersion <= currentVersion) {
+        // There are rare cases where the database is updated twice in quick succession, with the
+        // second update containing the same version as the first. In this case, we still need to
+        // apply the second set of updates. Skipping it would result in the database then containing
+        // references to entries that don't exist. This might be an engine issue, but accepting the
+        // second updates seems to be harmless, so we do that.
```
Comment (farmaazon) [inline app/gui2/src/stores/suggestionDatabase/index.ts:106]: Let's report that.

File: app/gui2/src/util/database/reactiveDb.ts [hunk 12]
```diff
@@ -112,10 +115,27 @@ export class ReactiveDb<K, V> extends ObservableV2<{
...
+   */
+  moveToLast(id: K) {
+    const value = this._internal.get(id)
+    if (value !== undefined) {
+      this._internal.delete(id)
+      this._internal.set(id, value)
+    }
+    // // Trigger deps that track iteration order of the entire map, but nothing else. Triggering a
+    // // delete operation without specifying any specific key does exactly that.
+    // if (value != null) trigger(this.internal, 'delete' as TriggerOpTypes.DELETE)
```
Comment (farmaazon) [inline app/gui2/src/util/database/reactiveDb.ts]: Commented code.

File: app/gui2/src/util/database/reactiveDb.ts [hunk 13]
```diff
@@ -151,8 +171,8 @@ export class ReactiveIndex<K, V, IK, IV> {
    * @param indexer - The indexer function defining how db keys and values map to index keys and values.
    */
   constructor(db: ReactiveDb<K, V>, indexer: Indexer<K, V, IK, IV>) {
-    this.forward = reactive(new Map())
-    this.reverse = reactive(new Map())
+    this.forward = reactive(map.create())
```
Comment (farmaazon) [inline app/gui2/src/util/database/reactiveDb.ts:169]: What are the gains of this change?

`lib0.map.create` does exactly `new Map()` https://github.com/dmonad/lib0/blob/main/map.js#L15

File: app/gui2/src/util/database/reactiveDb.ts [hunk 14]
```diff
@@ -218,6 +249,65 @@ export class ReactiveIndex<K, V, IK, IV> {
...
+ * Provides automatic indexing for a `ReactiveDb` instance. Allows only one value per key. It can be
+ * thought of as a collection of `computed` values per each key in the `ReactiveDb`. The mapping is
+ * automatically updated when any of its dependencies change, and is properly cleaned up when any
+ * key is removed from {@link ReactiveDb}. Only accessed keys are ever actually computed.
+ *
+ * @typeParam K - The key type of the ReactiveDb.
+ * @typeParam V - The value type of the ReactiveDb.
+ * @typeParam M - The type of a mapped value.
+ */
+export class ReactiveMapping<K, V, M> {
```
Comment (farmaazon) [inline app/gui2/src/util/database/reactiveDb.ts:278]: It is not "automatic indexing" because it does not index anything, i.e. do not make any lookup faster.

File: app/gui2/src/util/navigator.ts [hunk 15]
```diff
@@ -34,6 +34,21 @@ export function useNavigator(viewportNode: Ref<Element | undefined>) {
...
+      v.pos.x + v.size.x * (canvasPos.x / rect.size.x),
+      v.pos.y + v.size.y * (canvasPos.y / rect.size.y),
+    )
+    const size = new Vec2(
+      v.size.x * (clientRect.size.x / rect.size.x),
+      v.size.y * (clientRect.size.y / rect.size.y),
+    )
+    return new Rect(pos, size)
+  }
+
```
Comment (farmaazon) [inline app/gui2/src/util/navigator.ts:51]: Tests, please!

File: app/gui2/ydoc-server/languageServerSession.ts [hunk 16]
```diff
@@ -1,10 +1,17 @@
...
 import * as random from 'lib0/random'
 import * as Y from 'yjs'
-import { LanguageServer, computeTextChecksum } from '../shared/languageServer'
+import {
+  ErrorCode,
+  LanguageServer,
+  LsRpcError,
+  RemoteRpcError,
+  computeTextChecksum,
+} from '../shared/languageServer'
```
Comment (farmaazon) [inline app/gui2/ydoc-server/languageServerSession.ts]: Interesting. I see new imports, but no usage of those in this file.

**Review by farmaazon — APPROVED** [review]
Looks good, but I have so knowledge from GUI widgets implementation. It would be nice to add some overview for less lucky fellow programmers.

File: app/gui2/src/components/GraphEditor/GraphEdge.vue (hunk 1 above)
Comment (Frizi) [reply app/gui2/src/components/GraphEditor/GraphEdge.vue]: I tried to avoid using the graph store directly in the edge, but there probably is no point. We are already referencing the selection there anyway. So I'll just simplify that.

File: app/gui2/src/components/GraphEditor/widgets/WidgetPort.vue (hunk 6 above)
Comment (Frizi) [reply app/gui2/src/components/GraphEditor/widgets/WidgetPort.vue]: Currently it is, but I'm pretty sure this is not how we want it to be. Making all connectable things "ports" will simplify the hover/exprRect update handling.

File: app/gui2/src/providers/widgetTree.ts (hunk 8 above)
Comment (Frizi) [reply app/gui2/src/providers/widgetTree.ts]: This is all the constructor does - accepts and assigns one public property. Initially this class had more logic, but it turned out to be unnecessary for now. It might again have it later, or it will be removed once completely unnecessary.

File: app/gui2/src/util/database/reactiveDb.ts (hunk 13 above)
Comment (Frizi) [reply app/gui2/src/util/database/reactiveDb.ts:169]: Yes. But it is minified to smaller code (because there is no need for `new` keyword)`. That's the whole point of having it defined like that in lib0.

File: app/gui2/src/components/GraphEditor/widgets/WidgetPort.vue (hunk 6 above)
Comment (Frizi) [reply app/gui2/src/components/GraphEditor/widgets/WidgetPort.vue]: I've moved this logic to the actual port widget now.

File: app/gui2/src/providers/widgetTree.ts (hunk 8 above)
Comment (Frizi) [reply app/gui2/src/providers/widgetTree.ts]: Simplified this a bit by removing the class. There is no need for more logic there now.

**Review by somebody1234 — APPROVED** [review]
approving mostly to unblock this PR, may or may not do a proper review later

File: app/gui2/src/providers/interactionHandler.ts [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+
+  setCurrent(interaction: Interaction | undefined) {
+    if (interaction !== this.currentInteraction) {
+      this.currentInteraction?.cancel?.()
+      this.currentInteraction = interaction
+    }
+  }
+
+  /** Unset the current interaction, if it is the specified instance. */
+  end(interaction: Interaction) {
```
Comment (farmaazon) [inline app/gui2/src/providers/interactionHandler.ts:32]: Nitpick: I wonder what is the rule of which methods are documented, and which are not. For example, this particular method is documented enough by its code.

File: app/gui2/src/util/array.ts [hunk 18]
```diff
@@ -34,3 +34,8 @@ export function partitionPoint<T>(
   }
   return start
 }
+
+/** Index into an array using specified index. When the index is nullable, returns undefined. */
+export function tryGetIndex<T>(arr: T[], index: number | undefined | null): T | undefined {
```
Comment (farmaazon) [inline app/gui2/src/util/array.ts:39]: Why not `index: Opt<number>`?

File: app/gui2/src/util/rect.ts [hunk 19]
```diff
@@ -9,11 +9,19 @@ export class Rect {
     readonly size: Vec2,
   ) {}
 
+  static Zero = new Rect(Vec2.Zero, Vec2.Zero)
```
Comment (farmaazon) [inline app/gui2/src/util/rect.ts:12]: Shouldn't constants be rather `ZERO` than `Zero`?

**Review by farmaazon — APPROVED** [review]
Additional comments for recent changes.
